ReplaceChunk multiple line support

mispencer opened the pull request:
C# fixit support will need this.

mispencer committed:
ReplaceChunk multiple line support

Valloric commented on python/ycm/vimsupport.py:
```diff
@@ -465,17 +465,55 @@ def GetIntValue( variable ):
 # returns the delta (in characters) to the line length applied
-#
-# PreCondition: start and end are on the same line.
 def ReplaceChunk ( start, end, replacement_text, delta ):
```
This function is now seriously complex. Any chance of splitting it up? Either way, with this increase in complexity, this will need _very_ good test coverage before we can trust it to stay working.

mispencer replied:
I'll see what I can do.

However, how do we test something so closely tied to vim?

Valloric replied:
That's a good question. Look at the `base_test.py` file to see examples; it should be easy to understand & reproduce. It mostly boils down to using `MagicMock()` to mock out the vim interfaces.

mispencer commented:
I couldn't think of a good way to split up ReplaceChunk, so I settled for adding a few comments.

I couldn't figure out the vim mocking, so I added a optional parameter for ReplaceChunk for overriding the buffer for test mocking. I've added a bunch of tests.

puremourning commented:
I can't help wondering if we're totally doing this wrong. there's a little python in clang-format.py which uses [difflib](https://docs.python.org/2/library/difflib.html) like this:

```
    lines = stdout.split('\n')
    output = json.loads(lines[0])
    lines = lines[1:]
    sequence = difflib.SequenceMatcher(None, vim.current.buffer, lines)
    for op in reversed(sequence.get_opcodes()):
      if op[0] is not 'equal':
        vim.current.buffer[op[1]:op[2]] = lines[op[3]:op[4]]
    if output['IncompleteFormat']:
      print 'clang-format: incomplete (syntax errors)'
    vim.command('goto %d' % (output['Cursor'] + 1))
```

i mean, i haven't got that to work, but it is a lot less code that we're both written here!

mispencer commented:
It's been refactored.

I also added line_delta, since with multiple line clunk replacing, the line number for following clunks need to be adjusted as well.

puremourning commented:
Thanks!

mispencer commented:
```
  if replacement_lines_count > 1:
    if source_lines_count > 1:
      new_char_delta -= start_column
    else:
      new_char_delta -= end_column - char_delta
```

should be 

```
  if replacement_lines_count > 1:
    new_char_delta -= start_column
```

I'm not sure what I was thinking.

puremourning commented:
I spent 10 minutes looking at that, thinking the same thing. I checked the tests and it looked like you'd covered it so I left it as is. I will check it out.

puremourning commented:
well with that change, this test fails:

```

def ReplaceChunk_SingleToMultipleLines_test():
  result_buffer = [ "aAa", 
                    "aBa", 
                    "aCa" ]
  start, end = _BuildLocations( 2, 2, 2, 2 )
  ( line_offset, char_offset ) = vimsupport.ReplaceChunk( start, end, 'Eb\nbF',
                                                          0, 0, result_buffer )
  expected_buffer = [ "aAa", 
                      "aEb",
                      "bFBa",
                      "aCa" ]
  eq_( expected_buffer, result_buffer )
  eq_( line_offset, 1 )
  eq_( char_offset, 1 )

  # now make another change to the "2nd" line
  start, end = _BuildLocations( 2, 3, 2, 4 )
  ( new_line_offset, new_char_offset ) = vimsupport.ReplaceChunk( 
                                                           start, 
                                                           end, 
                                                           'cccc',
                                                           line_offset,
                                                           char_offset,
                                                           result_buffer )

  line_offset += new_line_offset
  char_offset += new_char_offset

  eq_( [ "aAa", "aEb", "bFBcccc", "aCa" ], result_buffer )
  eq_( line_offset, 1 )
  eq_( char_offset, 4 )
```

with

```

def ReplaceChunk_SingleToMultipleLines_test():
  result_buffer = [ "aAa", 
                    "aBa", 
                    "aCa" ]
  start, end = _BuildLocations( 2, 2, 2, 2 )
  ( line_offset, char_offset ) = vimsupport.ReplaceChunk( start, end, 'Eb\nbF',
                                                          0, 0, result_buffer )
  expected_buffer = [ "aAa", 
                      "aEb",
                      "bFBa",
                      "aCa" ]
  eq_( expected_buffer, result_buffer )
  eq_( line_offset, 1 )
  eq_( char_offset, 1 )

  # now make another change to the "2nd" line
  start, end = _BuildLocations( 2, 3, 2, 4 )
  ( new_line_offset, new_char_offset ) = vimsupport.ReplaceChunk( 
                                                           start, 
                                                           end, 
                                                           'cccc',
                                                           line_offset,
                                                           char_offset,
                                                           result_buffer )

  line_offset += new_line_offset
  char_offset += new_char_offset

  eq_( [ "aAa", "aEb", "bFBcccc", "aCa" ], result_buffer )
  eq_( line_offset, 1 )
  eq_( char_offset, 4 )
```

mispencer commented:
Hmm? That test passes for me.

puremourning commented:
sorry, i was being dumb. ignore that :)

puremourning commented:
ok pushed new tests and this fix
